### PR TITLE
correctly generate safe casts from Unit type

### PR DESCRIPTION
**Pull request by yole**
#KT-4494 fixed

### CODE REVIEW

File: compiler/backend/src/org/jetbrains/kotlin/codegen/ExpressionCodegen.java [hunk 1]
```diff
@@ -3833,6 +3833,9 @@ public Unit invoke(InstructionAdapter v) {
                                 v.mark(nonnull);
                             }
                         }
+                        else if (value.type == Type.VOID_TYPE) {
+                            v.aconst(null);
+                        }
                         else {
                             v.dup();
                             generateInstanceOfInstruction(rightType);
```
Comment (udalov) [inline compiler/backend/src/org/jetbrains/kotlin/codegen/ExpressionCodegen.java]: Similar code is present in `generateInstanceOf` which probably means `Unit is Int?` is also affected?

Comment (yole) [inline compiler/backend/src/org/jetbrains/kotlin/codegen/ExpressionCodegen.java]: That's not allowed by the frontend: "Incompatible types: kotlin.Int? and kotlin.Unit (3,31)"

**Comment by udalov** [comment]
Ok

**Comment by yole** [comment]
Merged manually

**Pull request closed without merging**